### PR TITLE
support for larger than 35 minute clock skew

### PR DESCRIPTION
**Pull request by ZappoMan**
This fixes the various places in the code where we treat clock skew as a 32bit int. The problem with that is that 32bit ints only allow for +/-35 minutes of clock skew. Although it seems unlikely that a clock would be off by more than 35 minutes, it could be possible if someone has for example misconfigured their time zone.

### CODE REVIEW

**Comment by ZappoMan** [comment]
Note: This is a client and server change... both can have the rollover error in them. It doesn't change the protocol between them. But larger than 35 minute clock skews one way or the other will only properly work if both client and server have been updated to the latest code.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31f9d592e0cf24d05965a8dfbac2241db51aa648 available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5244274ae4545bea58c66cff18de8d843d024505 available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

**Comment by ZappoMan** [comment]
Testing Notes:

**Smoke Test**
- Do a smoke test between this client and existing servers (assuming their clocks are close to correct) then everything should work normally.
- Do a smoke test between old clients and this server  (assuming their clocks are close to correct) then everything should work normally.

**Specific Tests:**
The goal is to do multi-user testing where the clock of one of the machines (client or server) is way out of sync.
- On two different machines, set one machine to be 1 hour behind the correct time. 
- Run this sandbox on one machine, and this client on a second machine. 
- On the client machine: check the Developer > Entities > Entities Statistics [more] -- verify that "Average Clock Skew" is approximately the difference you set the time to.
- Edit an entity in the client -- check the logs and make sure you don't see messages that say "ignoring unreasonable packet"
- Run a second client (can be on either machine, or a third machine) -- do edits on each client and verify that the other client sees the edits done by the second client. Test grabbing and editing with edit.js

- Repeat tests with: 1 minute ahead, 1 minute behind, 10 minutes ahead, 10 minutes behind, 1 hour ahead, 1 hour behind, 2 hours ahead, 2 hours behind, 1 day ahead, 1 day behind.

Note: You will need to restart the server and client on the machine you do time changes for.
Note: You may need to re-teleport to the server/domain or restart the client even on the machine that didn't change the clock. This PR doesn't attempt to handle changes in server time on the fly.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ddf5b0f3160918737408a2eb2f0dcf047becf1af available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: f9c4959578b313af05d3ffef27b49d53f3bf8feb available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

File: libraries/networking/src/Node.cpp [hunk 1]
```diff
@@ -73,9 +73,9 @@ void Node::setType(char type) {
     _symmetricSocket.setObjectName(typeString);
 }
 
-void Node::updateClockSkewUsec(int clockSkewSample) {
-    _clockSkewMovingPercentile.updatePercentile((float)clockSkewSample);
-    _clockSkewUsec = (int)_clockSkewMovingPercentile.getValueAtPercentile();
+void Node::updateClockSkewUsec(qint64 clockSkewSample) {
+    _clockSkewMovingPercentile.updatePercentile((double)clockSkewSample);
```
Comment (zzmp) [inline libraries/networking/src/Node.cpp]: updatePercentile takes a float, so you're still going to be doing that conversion.

Comment (zzmp) [inline libraries/networking/src/Node.cpp]: You changed this to take a qint64, so you should not be casting it.

File: libraries/shared/src/MovingPercentile.h [hunk 2]
```diff
@@ -18,19 +18,21 @@ class MovingPercentile {
 public:
     MovingPercentile(int numSamples, float percentile = 0.5f);
 
-    void updatePercentile(float sample);
-    float getValueAtPercentile() const { return _valueAtPercentile; }
+    // FIXME - this class is only currently used in calculating the clockSkew, which is a signed 64bit int, not a double
+    // I am somewhat tempted to make this a type sensitive template and/or swith to using qint64's internally
```
Comment (zzmp) [inline libraries/shared/src/MovingPercentile.h]: switch

File: libraries/networking/src/Node.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/networking/src/Node.cpp]: let me fix that - this was a half way change

File: libraries/shared/src/MovingPercentile.h [hunk 3]
```diff
@@ -18,19 +18,21 @@ class MovingPercentile {
 public:
     MovingPercentile(int numSamples, float percentile = 0.5f);
 
-    void updatePercentile(float sample);
-    float getValueAtPercentile() const { return _valueAtPercentile; }
+    // FIXME - this class is only currently used in calculating the clockSkew, which is a signed 64bit int, not a double
```
Comment (zzmp) [inline libraries/shared/src/MovingPercentile.h]: Template! But this is fine for now.

File: libraries/shared/src/SharedUtil.cpp [hunk 4]
```diff
@@ -621,16 +621,65 @@ void debug::checkDeadBeef(void* memoryVoid, int size) {
     assert(memcmp((unsigned char*)memoryVoid, DEADBEEF, std::min(size, DEADBEEF_SIZE)) != 0);
 }
 
+QString formatUsecTime(quint64 usecs, int prec) {
```
Comment (zzmp) [inline libraries/shared/src/SharedUtil.cpp]: Can you template this to avoid rewriting it?

File: libraries/shared/src/SharedUtil.cpp [hunk 5]
```diff
@@ -43,8 +43,8 @@
 #include "OctalCode.h"
 #include "SharedLogging.h"
 
-static int usecTimestampNowAdjust = 0; // in usec
-void usecTimestampNowForceClockSkew(int clockSkew) {
+static qint64 usecTimestampNowAdjust = 0; // in usec
```
Comment (Atlante45) [inline libraries/shared/src/SharedUtil.cpp]: Kinda mind-blowing that simply doubling the bit size gets you from ~35min to ~300,000 years.

Comment (ZappoMan) [inline libraries/shared/src/SharedUtil.cpp]: I think it's only 300 years...

File: libraries/networking/src/NodeList.cpp [hunk 6]
```diff
@@ -158,18 +158,19 @@ void NodeList::timePingReply(ReceivedMessage& message, const SharedNodePointer&
     sendingNode->setPingMs(pingTime / 1000);
     sendingNode->updateClockSkewUsec(clockSkew);
 
-    const bool wantDebug = false;
+    const bool wantDebug = true;
```
Comment (zzmp) [inline libraries/networking/src/NodeList.cpp]: Shouldn't this be false?

File: libraries/octree/src/OctreeSceneStats.cpp [hunk 7]
```diff
@@ -747,7 +747,7 @@ const char* OctreeSceneStats::getItemValue(Item item) {
 }
 
 void OctreeSceneStats::trackIncomingOctreePacket(ReceivedMessage& message, bool wasStatsPacket, qint64 nodeClockSkewUsec) {
-    const bool wantExtraDebugging = false;
+    const bool wantExtraDebugging = true;
```
Comment (zzmp) [inline libraries/octree/src/OctreeSceneStats.cpp]: Shouldn't this be false?

File: libraries/shared/src/SharedUtil.cpp (hunk 5 above)
Comment (Atlante45) [inline libraries/shared/src/SharedUtil.cpp]: Nope, 292,471.208677536 years to be exact.
Unless I messed up my formula "2^64 / (2 * 1000 * 1000 * 60 * 60 * 24 * 365)", don't think I did.

File: assignment-client/src/octree/OctreeServer.cpp [hunk 8]
```diff
@@ -952,6 +952,25 @@ bool OctreeServer::readOptionInt(const QString& optionName, const QJsonObject& s
...
 }
 
+bool OctreeServer::readOptionInt64(const QString& optionName, const QJsonObject& settingsSectionObject, qint64& result) {
+    bool optionAvailable = false;
+    QString argName = "--" + optionName;
+    const char* argValue = getCmdOption(_argc, _argv, qPrintable(argName));
+    if (argValue) {
+        optionAvailable = true;
+        result = atoll(argValue);
+        qDebug() << "From payload arguments: " << qPrintable(argName) << ":" << result;
```
Comment (sethalves) [inline assignment-client/src/octree/OctreeServer.cpp]: did you intend to leave these qDebugs?

File: interface/src/main.cpp [hunk 9]
```diff
@@ -127,9 +127,9 @@ int main(int argc, const char* argv[]) {
     const char* CLOCK_SKEW = "--clockSkew";
     const char* clockSkewOption = getCmdOption(argc, argv, CLOCK_SKEW);
     if (clockSkewOption) {
-        int clockSkew = atoi(clockSkewOption);
+        quint64 clockSkew = atoll(clockSkewOption);
```
Comment (sethalves) [inline interface/src/main.cpp]: why a quint64 rather than a qint64 ?

File: libraries/networking/src/Node.cpp [hunk 10]
```diff
@@ -73,9 +73,9 @@ void Node::setType(char type) {
...
+    _clockSkewUsec = (quint64)_clockSkewMovingPercentile.getValueAtPercentile();
```
Comment (sethalves) [inline libraries/networking/src/Node.cpp]: qint64 rather than quint64?

File: assignment-client/src/octree/OctreeServer.cpp (hunk 8 above)
Comment (ZappoMan) [inline assignment-client/src/octree/OctreeServer.cpp]: yes - these are in the other version of the function... but I just noticed that VS messed up the brace placement - so I will fix that

File: libraries/shared/src/SharedUtil.cpp [hunk 11]
```diff
@@ -621,72 +624,75 @@ void debug::checkDeadBeef(void* memoryVoid, int size) {
...
-    } else if (usecs > USECS_PER_SECOND || usecs < -USECS_PER_SECOND) {
-        result = QString::number(usecs / USECS_PER_SECOND) + 's';
-    } else if (usecs > USECS_PER_MSEC || usecs < -USECS_PER_MSEC) {
-        result = QString::number(usecs / USECS_PER_MSEC) + "ms";
-    } else {
-        result = QString::number(usecs) + "us";
-    }
-    return result;
+
+QString formatUsecTime(quint64 usecs) {
```
Comment (zzmp) [inline libraries/shared/src/SharedUtil.cpp]: Did you do it this way to avoid putting it in the header file?

Comment (ZappoMan) [inline libraries/shared/src/SharedUtil.cpp]: I didn't want to include the template in the header file, and I didn't want all the callers to change to having to use the template form.

File: libraries/shared/src/SharedUtil.cpp (hunk 5 above)
Comment (ZappoMan) [inline libraries/shared/src/SharedUtil.cpp]: actually it should be 2^63 - 1... but yeah... that's a lot of clock skew!

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9ae788818f4f6a9e9b3780cc34b0cb10705b42f available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 72c81e237809d41a008a6bde06ad7ddd887704f7 available at: https://gist.github.com/57a82f67f531699c2a870304ed0e9bc4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7727.dmg

File: libraries/shared/src/SharedUtil.cpp (hunk 5 above)
Comment (Atlante45) [inline libraries/shared/src/SharedUtil.cpp]: I divided by 2, but I'll give you the -1 ;)

**Comment by OmegaHeron** [comment]
Phase 1 - General smoke test.
This PR to old servers
Heron (RC6) - good
O2Tech (Latest devel) - good
Playa - good
Cellscience - good
Forum - good

RC6 Interface (Windows 7) to this PR server (Linux over inet) - good
This PR Interface to this PR server - good

This PR Interface general smoke tests (visit playa, cellscience, forum - place built in entities in my domain - make sure ATP works) - good

Phase 2 will be a fair bit longer to come.

**Comment by OmegaHeron** [comment]
Specific test 1
Interface system Windows 7 - Windows machine one hour behind correct time - Server Linux this PR with correct time.

Time difference matches 1 hour offset properly.  I show negative offset of almost precisely 1 hour in octree stats window.  Log **does not** output ignoring unreasonable packet as it does in other releases.

More to come

**Comment by OmegaHeron** [comment]
Also - go go NTP on both sides

**Average Clock Skew: 0 msecs [-943 usecs]**

**Comment by ZappoMan** [comment]
@OmegaHeron - btw - you don't have to do every combination of +1/-1 day/hour/minute/etc... but just a good sampling...

**Comment by OmegaHeron** [comment]
Yeah I mixed it up some.  Have thunderstorms now and had to unplug for a bit, will see if I can draft a helper in a bit to finish once storms clear out.

**Comment by OmegaHeron** [comment]
Was able to run final test on LAN - 3 computers, clients on each machine with varying incorrect local time.  No "unauthorized" messages in log - I could see my edits across both AVs.  Would rather have tested this over internet, but storms have made things tricky here this evening.  I think I covered everything.  QA Pass

**Pull request merged**